cardinal/events/events.go

```diff
@@ -9,10 +10,84 @@ import (
 	"github.com/go-openapi/runtime/middleware"
 	"github.com/gorilla/websocket"
 	"github.com/rs/zerolog/log"
+	ecslog "pkg.world.dev/world-engine/cardinal/ecs/log"
 )
 
-func CreateEventHub() *EventHub {
-	res := EventHub{
+type EventHub interface {
+	EmitEvent(event *Event)
+	FlushEvents()
+	ShutdownEventHub()
+	Run()
+}
+
+type LoggingEventHub struct {
+	logger     *ecslog.Logger
+	eventQueue []*Event
+	running    atomic.Bool
+	broadcast  chan *Event
+	shutdown   chan bool
+	flush      chan bool
+}
+
+func (eh *LoggingEventHub) EmitEvent(event *Event) {
+	eh.broadcast <- event
+}
+
+func (eh *LoggingEventHub) FlushEvents() {
+	eh.flush <- true
+}
+
+func (eh *LoggingEventHub) Run() {
+	if eh.running.Load() {
+		return
+	}
+	eh.running.Store(true)
+	for eh.running.Load() {
+		select {
+		case event := <-eh.broadcast:
+			eh.eventQueue = append(eh.eventQueue, event)
+		case <-eh.flush:
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				acc := 0
+				for _, event := range eh.eventQueue {
+					eh.logger.Info().Msg(event.Message)
+					acc += 1
+				}
+				fmt.Printf("called %d times\n", acc)
+			}() //a goroutine is not technically necessary here but this imitates the websocket eventhub as much as possible.
+			wg.Wait()
+			eh.eventQueue = eh.eventQueue[:0]
+		case <-eh.shutdown:
+			eh.running.Store(false)
+		}
+	}
+}
+
+func (eh *LoggingEventHub) ShutdownEventHub() {
+	eh.shutdown <- true
+}
+
+func CreateLoggingEventHub(logger *ecslog.Logger) *LoggingEventHub {
+	res := LoggingEventHub{
+		eventQueue: make([]*Event, 0),
+		running:    atomic.Bool{},
+		broadcast:  make(chan *Event),
+		shutdown:   make(chan bool),
+		flush:      make(chan bool),
+		logger:     logger,
+	}
+	res.running.Store(false)
+	go func() {
+		res.Run()
+	}()
+	return &res
+}
+
+func CreateWebSocketEventHub() *WebSocketEventHub {
```

Comment:
The WebSocketEventHub is a thing that must be shared with the server and ecs.World

The LoggingEventHub is exclusive to ecs.World. 

Thus ecs.World can use a shared Interface. However, server must use socket related functions such as ServeHttp.

Reply:
In short, this can't be done.

Reply:
RegisterConnection. It doesn't make sense to register a connection on loggerevent hub since all it does is log.

Reply:
ahhh i definitely misunderstood the design before -- i don't think Cardinal should be handling multiple connections. the job of relaying to multiple clients is deferred to the relayer (nakama). 

but thats my bad, i should have caught this previously and flagged it down.

Reply:
Don't worry. Relaying is still handled by nakama. 

This is just a websocket server. Like the http server it CAN handle multiple connections. 

For http although nakama is the only client to cardinal, the cardinal http server is handling multiple connections per call from the user side already via http. 

This is a useable model for events as well, especially if nakama uses web sockets for events. 

It depends on my options but making eventHub a proper websocket server rather than a one off that only accepts one connection allows me to anticipate every eventuality including handling a singleton connection and the one where we one day get rid of nakama and one where there are multiple connections between nakama and cardinal. 

Either way I also put a lot of work into this. Back tracking will be a pain. It's your call. Let me know what you want.

Reply:
ok got it. scratch that then. going back to the other thing, maybe we can just add `RegisterConnection` to the interface, and have the logger impl do nothing for it 🤷

Reply:
Also this issue is orthoganol to the thread at hand. The webSocketEventhub must register a connection whether that's one or multiple is a seperate issue. 

Actually I can just have the register function throw an error if the connections exceed more than one if you want me to cripple the websocket server. That I believe will give you the functionality you're looking for without too dramatic of a change.

Reply:
> ok got it. scratch that then. going back to the other thing, maybe we can just add `RegisterConnection` to the interface, and have the logger impl do nothing for it 🤷

I don't agree with this. But whatever. I can do it.